### DISCUSSION TITLE
Overriding the header for wrapped baseQuery

### DISCUSSION THREAD

**Discussion by checkDaResume** [reduxjs/redux-toolkit#2054] [Q&A] [answered]
Hi all,

I am using the a wrapped basequery to re-authenticate a user when the active token expires. However, the default baseQuery request carries a header which contains the expired token. So in the wrapped basequery my taught is to supply the `refresh` token in place of this expired token. However, I don't seem to know how to correctly pass `headers` in basequery on the re-auth request. At the moment if the token expires, it just logs-out which makes sense. However, I will only want this to happen when both `access` and `refresh` tokens are expired. Kindly see the conditional statement that I am using to execute the re-auth below:

```js

    if (result.error && result.error.status === 401){
        const refreshResult = await baseQuery(
            {url: 'auth/refresh', 
            headers: `Authorization Bearer ${user.refresh}` // <-- this is WRONG but I wanted to show one of my attempts
            }, api, extraOptions)
        if(refreshResult.data){
            api.dispatch(getuser()) // <-- I am not passing any arguments to getuser() but I will.
            result = await baseQuery(args, api, extraOptions)
        }else{
            api.dispatch(removeuser())
        }
    }

```

**Comment by phryneas** [comment, marked as answer]
```js
headers: {
  authorization: `Bearer ${user.refresh}`
}
```

as if you are using [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/fetch#parameters) :)

**Reply by checkDaResume** [reply]
Ahaaa. :-D thanks so much @phryneas. You always come through for me. I will implement this right away. Cheers.

**Reply by checkDaResume** [reply]
I having a little trouble getting the server to receive the `refresh` token. I did some troubleshooting and it seems the issue still has to do with the headers being sent in the wrapped baseQuery. I know this because if I send the `refresh` token in the headers directly on the default baseQuery the server receives it just fine. However, if I try to override the  expired `access` token on the default baseQuery it just goes ahead and sends the expired `access` token on the default baseQuery. I think I am missing something here. For completeness this is how I have my condition query setup:

```js

    if (result.error && result.error.status === 401){
        const refreshResult = await baseQuery(
            {url: 'auth/refresh/', 
            headers: { authorization : `Bearer ${user.refresh}`},  // <--- I also used {'Authorization' : `Bearer ${user.refresh}`} did'nt work either
            }, api, extraOptions)
            console.log(refreshResult)
        if(refreshResult.data){
            api.dispatch(adduser(refreshResult.data))
            result = await baseQuery(args, api, extraOptions)
        }else{
            api.dispatch(removeuser())
        }
    }

```

Thanks in advance.

**Reply by phryneas** [reply]
Are you maybe blindly overwriting it in your `prepareHeaders` call?

Data flow is as following
* baseQuery default `header`
* endpoint `header`
* `prepareHeaders` can make changes

So if you always set it in `prepareHeaders`, it will overwrite it. Use some logic there to see if it is already set.

**Reply by checkDaResume** [reply]
@phryneas That was actually it. I didn't realize that the `prepareHeaders` came last in the chain. I have now created a boolean in `prepareHeaders` which checks if the `Authorization` key is set and it works as expected. Thanks so much sir.
